src/plus/gk/serverConnection.ts

```diff
@@ -169,28 +192,180 @@ export class ServerConnection implements Disposable {
+		return new ProviderRequestRateLimitError(ex, token, resetAt);
+	}
+
+	private async handleUnsuccessfulResponse(rsp: Response, scope: LogScope | undefined): Promise<void> {
+		// Too Many Requests
+		if (rsp.status == 429) {
+			this.trackRequestException();
+			return;
+		}
+
+		// Forbidden
+		if (rsp.status == 403) {
+			if (rsp.statusText.includes('rate limit')) {
+				this.trackRequestException();
+			}
+			return;
+		}
+
+		// Internal Server Error
+		if (rsp.status == 500) {
+			this.trackRequestException();
+			void showIntegrationRequestFailed500WarningMessage(
+				'GitKraken failed to respond and might be experiencing issues. Please visit the [GitKraken status page](https://cloud.gitkrakenstatus.com) for more information.',
+			);
+			return;
+		}
+
+		const content = await rsp.text();
+
+		// Bad Gateway
+		if (rsp.status == 502) {
+			Logger.error(`GitKraken request failed: ${content} (${rsp.statusText})`, scope);
+			if (content.includes('timeout')) {
+				this.trackRequestException();
+				void showIntegrationRequestTimedOutWarningMessage('GitKraken');
+			}
+			return;
+		}
+
+		// Service Unavailable
+		if (rsp.status == 503) {
+			Logger.error(`GitKraken request failed: ${content} (${rsp.statusText})`, scope);
+			this.trackRequestException();
+			void showIntegrationRequestFailed500WarningMessage(
+				'GitKraken failed to respond and might be experiencing issues. Please visit the [GitKraken status page](https://cloud.gitkrakenstatus.com) for more information.',
+			);
+			return;
+		}
+
+		if (rsp.status >= 400 && rsp.status < 500) {
+			return;
+		}
+
+		if (Logger.isDebugging) {
+			void window.showErrorMessage(`GitKraken request failed: ${content} (${rsp.statusText})`);
+		}
+	}
+
+	private handleRequestError(
+		token: string | undefined,
+		ex: RequestError | (Error & { name: 'AbortError' }),
+		scope: LogScope | undefined,
+	): void {
+		if (ex instanceof CancellationError) throw ex;
+		if (ex.name === 'AbortError') throw new CancellationError(ex);
+
+		switch (ex.status) {
+			case 404: // Not found
+				throw new ProviderRequestNotFoundError(ex);
+			case 410: // Gone
+				throw new ProviderRequestGoneError(ex);
+			case 422: // Unprocessable Entity
+				throw new ProviderRequestUnprocessableEntityError(ex);
+			case 401: // Unauthorized
+				throw new AuthenticationError('gitkraken', AuthenticationErrorReason.Unauthorized, ex);
+			case 429: //Too Many Requests
+				this.trackRequestException();
+				throw this.buildProviderRequestRateLimitError(token, ex);
+			case 403: // Forbidden
+				if (ex.message.includes('rate limit')) {
+					this.trackRequestException();
+					throw this.buildProviderRequestRateLimitError(token, ex);
+				}
+				throw new AuthenticationError('gitkraken', AuthenticationErrorReason.Forbidden, ex);
+			case 500: // Internal Server Error
+				Logger.error(ex, scope);
+				if (ex.response != null) {
+					this.trackRequestException();
+					void showIntegrationRequestFailed500WarningMessage(
```

Comment:
This function, and similar ones used in other cases, is specifically designed for integration messages. Its key tracking and other handling are specific to integration/provider requests. We need a more generic version of these "show message" functions that's scoped to GK requests.

Reply:
@axosoft-ramint 

Question. So, these functions use the general method `showMessage(...)` and the difference is the `message` text itself and the `suppression-key` that lets us a specific message type.

So, do I understand right that I should keep using this `showMessage` function by providing different suppression keys. such as
```
showGkRequestFailed500WarningMessage(message) {
   showMessage('error', message, 'showGkRequestFailed500WarningMessage', ...)
}
```

So, I shouldn't create a different version of the general `showMessage()` function, should I?

Reply:
done. Is it what you mean?

https://github.com/gitkraken/vscode-gitlens/commit/49083abe2976b6f3e97d3d3150941c3ee83f5c6c

Reply:
@sergeibbb Yes that looks right, thanks!

Reply:
@axosoft-ramint 

Ok. So, now I'm testing it in its final state and then it can be considered for being merged.